en/middleware/mavlink.md

```diff
@@ -2,200 +2,419 @@
+This tutorial demonstrates how to stream a uORB message as a MAVLink message, and applies to both standard and custom messages.
 
-Add the headers of the MAVLink and uORB messages to
-[mavlink_messages.cpp](https://github.com/PX4/PX4-Autopilot/blob/main/src/modules/mavlink/mavlink_messages.cpp)
+### Preconditions
 
-```C
-#include <uORB/topics/ca_trajectory.h>
-#include <v2.0/custom_messages/mavlink.h>
+Generally you will already have a [uORB](../middleware/uorb.md) message that contains information you'd like to stream and a definition of a MAVLink message that you'd like to stream it with.
+
+For this example we're going to assume that you want to stream the (existing) [BatteryStatus](../msg_docs/BatteryStatus.md) uORB message to a new MAVLink battery status message, which we will name `BATTERY_STATUS_DEMO`.
+
+Copy this `BATTERY_STATUS_DEMO` message into the message section of `development.xml` in your PX4 source code, which will be located at: `\src\modules\mavlink\mavlink\message_definitions\v1.0\development.xml`.
+
+```xml
+    <message id="11514" name="BATTERY_STATUS_DEMO">
+      <description>Simple demo battery.</description>
+      <field type="uint8_t" name="id" instance="true">Battery ID</field>
+      <field type="int16_t" name="temperature" units="cdegC" invalid="INT16_MAX">Temperature of the whole battery pack (not internal electronics). INT16_MAX field not provided.</field>
+      <field type="uint8_t" name="percent_remaining" units="%" invalid="UINT8_MAX">Remaining battery energy. Values: [0-100], UINT8_MAX: field not provided.</field>
+    </message>
 ```
 
-Create a new class in [mavlink_messages.cpp](https://github.com/PX4/PX4-Autopilot/blob/main/src/modules/mavlink/mavlink_messages.cpp#L2193)
+:::note
+Note that this is a cut-down version of the not-yet-implemented [BATTERY_STATUS_V2](https://mavlink.io/en/messages/development.html#BATTERY_STATUS_V2) message with randomly chosen unused id of `11514`.
+Here we've put the message in `development.xml`, which is fine for testing and if the message is intended to eventually be part of the standard message set, but you might also put a [custom message](#custom-mavlink-messages) in its own dialect file.
+:::
+
+Build PX4 for SITL and confirm that the associated message is generated in `/build/px4_sitl_default/mavlink/common/mavlink_msg_battery_status_demo.h`.
 
-```C
-class MavlinkStreamCaTrajectory : public MavlinkStream
+Because `BatteryStatus` already exists you will not need to do anything to create or build it.
+
+### Define the Streaming Class
+
+First create a file named `BATTERY_STATUS_DEMO.hpp` for your streaming class (named after the message to stream) inside the [/src/modules/mavlink/streams](https://github.com/PX4/PX4-Autopilot/tree/main/src/modules/mavlink/streams) directory.
+
+Add the headers for the MAVLink and uORB messages to the top of the file:
+
+```cpp
+#include <uORB/topics/battery_status.h>
+#include <v2.0/development/mavlink.h>
```

Comment:
You shouldn't need to import mavlink here right? I.e. headers are imported at PX4 level based on the config variable, and by this point you should get the correct selection?

Reply:
Yes exactly